test/validation/api/timer/timer.c

```diff
@@ -2170,6 +2197,238 @@ static void timer_test_sched_all(void)
+		 timer_param.periodic.base_freq_hz.integer,
+		 timer_param.periodic.base_freq_hz.numer,
+		 timer_param.periodic.base_freq_hz.denom, freq);
+	ODPH_DBG("  Max multiplier:    %" PRIu64 "\n", timer_param.periodic.max_multiplier);
+	ODPH_DBG("Capabilities:\n");
+	ODPH_DBG("  Max multiplier:    %" PRIu64 " (with %f hz)\n",
+		 periodic_capa.max_multiplier, freq);
+	ODPH_DBG("  Max resolution:    %" PRIu64 " ns (with %f hz)\n", periodic_capa.res_ns, freq);
+	ODPH_DBG("  Min base freq:     %f hz\n", min_freq);
+	ODPH_DBG("  Max base freq:     %f hz\n", max_freq);
+
+	timer_pool = odp_timer_pool_create("periodic_timer", &timer_param);
+	CU_ASSERT_FATAL(timer_pool != ODP_TIMER_POOL_INVALID);
+
+	odp_timer_pool_start();
+
+	odp_pool_param_init(&pool_param);
+	pool_param.type    = ODP_POOL_TIMEOUT;
+	pool_param.tmo.num = 1;
+
+	pool = odp_pool_create("timeout_pool", &pool_param);
+	CU_ASSERT_FATAL(pool != ODP_POOL_INVALID);
+
+	odp_queue_param_init(&queue_param);
+	if (queue_type == ODP_QUEUE_TYPE_SCHED) {
+		queue_param.type       = ODP_QUEUE_TYPE_SCHED;
+		queue_param.sched.sync = ODP_SCHED_SYNC_ATOMIC;
+	}
+
+	queue = odp_queue_create("timeout_queue", &queue_param);
+	CU_ASSERT_FATAL(queue != ODP_QUEUE_INVALID);
+
+	tmo = odp_timeout_alloc(pool);
+	ev  = odp_timeout_to_event(tmo);
+	CU_ASSERT_FATAL(ev != ODP_EVENT_INVALID);
+
+	timer = odp_timer_alloc(timer_pool, queue, user_ctx);
+	CU_ASSERT_FATAL(timer != ODP_TIMER_INVALID);
+
+	memset(&start_param, 0, sizeof(odp_timer_periodic_start_t));
+	cur_tick = odp_timer_current_tick(timer_pool);
+	tick = cur_tick + odp_timer_ns_to_tick(timer_pool, period_ns / 2);
+
+	start_param.first_tick = tick;
+	start_param.freq_multiplier = multiplier;
+	start_param.tmo_ev = ev;
+
+	ODPH_DBG("Periodic timer start:\n");
+	ODPH_DBG("  Current tick:    %" PRIu64 "\n", cur_tick);
+	ODPH_DBG("  First tick:      %" PRIu64 "\n", start_param.first_tick);
+	ODPH_DBG("  Multiplier:      %" PRIu64 "\n", start_param.freq_multiplier);
+	ODPH_DBG("Test duration ns:  %" PRIu64 "\n", duration_ns);
+
+	ret = odp_timer_periodic_start(timer, &start_param);
+
+	if (ret == ODP_TIMER_TOO_NEAR)
+		ODPH_ERR("First tick too near\n");
+	else if (ret == ODP_TIMER_TOO_FAR)
+		ODPH_ERR("First tick too far\n");
+	else if (ret == ODP_TIMER_FAIL)
+		ODPH_ERR("Periodic timer start failed\n");
+
+	CU_ASSERT_FATAL(ret == ODP_TIMER_SUCCESS);
+
+	t1 = odp_time_local();
+
+	/* Wait for timeouts. Make sure that scheduler context is not hold when
+	 * exiting the loop. */
+	while (1) {
+		if (queue_type == ODP_QUEUE_TYPE_SCHED)
+			ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+		else
+			ev = odp_queue_deq(queue);
+
+		if (ev == ODP_EVENT_INVALID) {
+			t2 = odp_time_local();
+			diff_ns = odp_time_diff_ns(t2, t1);
+			if (diff_ns > (10 * duration_ns))
+				break;
+
+			if (num_tmo >= num)
+				break;
+
+			continue;
+		}
+
+		CU_ASSERT(odp_event_type(ev) == ODP_EVENT_TIMEOUT);
+		CU_ASSERT(odp_timer_ack(timer, ev) == 0);
```

Comment:
Perhaps it's better to use `CU_ASSERT_FATAL()` to check that the event is of type `ODP_EVENT_TIMEOUT ` or just free the event without calling `odp_timer_ack()`.

Reply:
Fixed in v5